se3cnn/SO3.py

```diff
@@ -478,32 +476,43 @@ def spherical_harmonics_coeff_to_sphere(coeff, alpha, beta):
     return s
 
 
-
 ################################################################################
 # Linear algebra
 ################################################################################
 
-def get_matrix_kernel(A, eps=1e-10):
-    """
-    Compute an orthonormal basis of the kernel (x_1, x_2, ...)
-    A x_i = 0
-    scalar_product(x_i, x_j) = delta_ij
-
-    :param A: matrix
-    :return: matrix where each row is a basis vector of the kernel of A
-    """
-    _u, s, v = torch.svd(A)
-
-    # A = u @ torch.diag(s) @ v.t()
-    kernel = v.t()[s < eps]
-    return kernel
-
-
-def get_matrices_kernel(As, eps=1e-10):
-    """
-    Computes the commun kernel of all the As matrices
-    """
-    return get_matrix_kernel(torch.cat(As, dim=0), eps)
```

Comment:
I was concatenating the matrices together to get their null space (kernel is the same as null space).
I see that you are adding the matrices together instead.
I guess that the probability that you grow the null space by adding matrices together is small but still I prefer if we can make a more correct computation if we can.

for instance `[[1,0],[0,1]]` and `[[1,0],[0,-1]]` have no null space. But their sum has a dim 1 null space.

Reply:
concatenated `D` matrices form (block) matrix `A = [D1 D2 D3 D4 D5 ...]`, singular vectors of it are the same as eigenvectors of `B=A^T @ A`, as we have two compatible block matrices multiplied, we can expand expression block by block, which leads to the summation as in code

Reply:
for your example it will be 
```[[1, 0], [0, 1]]^T @ [[1, 0], [0, 1]] + [[1, 0], [0, -1]]^T @ [[1, 0], [0, -1]] = [[1, 0], [0, 1]] + [[1, 0], [1, 0]] = 2*[[1, 0], [0, 1]]
``` - no null space

Reply:
for your example it will be 
`
[[1,0],[0,1]]^T @ [[1,0],[0,1]] + [[1,0],[0,-1]]^T @ [[1,0],[0,-1]] = 2 * [[1,0],[0,1]]
`
- no null space

Reply:
clever, could we put this trick into `get_matrices_kernel` ?

Reply:
not really, as it uses svd we never have access to `A^T @ A` even if internals construct it (which is not necessary happens, depends on underlying algorithm)

Reply:
I don't understand the problem.
`get_matrices_kernel(Ds)` gets a list of matrices. it just has to return `get_matrix_kernel(D1.T @ D1 + D2.T @ D2 + D3.T @ D3)`

Reply:
if we do so, then we have to use eigh or similar,
and if you mean why not keep original function names/structure, while changing only few lines inside, the reason is that in python one cannot free memory of an object passes to the function within said function, and that difference appeared quite substantial.

Reply:
yes, let use eigh or similar!
Optimization makes the code more ugly :-( 
Do we really need this memory ?^^

Reply:
I personally do not, because I do not use higher order CG, 
but it makes a difference whether one can fully calculate L=9+ within 16+2 GB of RAM

Reply:
it makes difference whether one can fully calculate L=9 and higher within 16+2 GB of RAM

Reply:
ok then